src/psyclone/psyir/nodes/node.py

```diff
@@ -1107,15 +1109,49 @@ def ancestor(self, my_type, excluding=None, include_self=False,
                 f"The 'limit' argument to ancestor() must be an instance of "
                 f"Node but got '{type(limit).__name__}'")
 
+        # If we need to find a shared ancestor, then find a starting ancestor
+        # for the sharing node.
+        shared_ancestor = None
+        if shared_with is not None:
+            shared_ancestor = shared_with.ancestor(
+                    my_type, excluding=excluding,
+                    include_self=include_self, limit=limit)
+
         while myparent is not None:
             if isinstance(myparent, my_type):
                 if not (excluding and isinstance(myparent, excludes)):
+                    # If this is a valid instance but not the same as for
+                    # the shared_with node, we do logic afterwards to continue
+                    # searching upwards, as we could be either higher or
+                    # lower than the shared_ancestor found previously.
+                    if shared_ancestor and shared_ancestor is not myparent:
```

Comment:
Yeah

Reply:
We actually can't do this, because we need to only do it if `shared_ancestor is not None`, not if `myparent is not None` (the latter statement is always true if `shared_ancestor` is None, and the former statement prevents this falling through if we don't care about a shared ancestor).